### ISSUE TITLE
Questions about project Target and Roadmaps

### ISSUE THREAD

**Issue by WeiZhang555** [rficcaglia/guardian#2]
Hi Robert, this is Wei Zhang from Alibaba, I'm very interested in this project, but I still have some question about the project target and roadmaps, I tried to ping you on slack this morning but you're not online 😄 

I read all the docs and some issues, but I'm still not very clear about the project due to I'm a newbie to formal verification. 

What's the project's plan? Does it aim to formal verify Kubernetes' security design with TLA+ etc? Or does it want to verify users' policies with Z3 or Rego? What's the plan?

And what's the current status? Does it have a clear roadmap?

**Comment by rficcaglia** [comment]
Hi! Great questions!

I would agree that the project is very unclear and abstract at the moment.
That was somewhat intentional since this was initially a repo for my own
notes.  However, I have resolved to try and publish a proof of concept.

I will try to clarify the README and RBAC markdown at least this first
goal:  to create a prototype tool for Kubernetes RBAC that is similar in
functionality to AWS Zelkova.  This would allow kubernetes users to verify
that their RBAC configuration admits no unexpected access to resources.
This is very similar to AWS's IAM advisory tool, and its S3 bucket access
checks. I believe both of these are powered by Zelkova.  I believe
Microsoft uses similar tools for Azure.

If there were sufficient interest, the second tool would be similar to AWS
Tiros and Microsoft SecGuru. That would try to formally verify Kubernetes
NetworkPolicy.

To answer your specific questions:

> Does it aim to formal verify Kubernetes' policy rules?
Yes, starting with RBAC, then NetworkPolicy.

>Or does it want to verify users' policies with Z3
Yes, I am initially using Z3 as a solver, but there is no hard requirement
to do so.  There are many other solvers available.  Z3 just seems to be
well documented and easy to start with.  I think Zelkova also uses it.

>or Rego?
Rego is itself a datalog-like policy language.  As such, it overlaps with
many of the ideas (Horn clause formulas, definite logic, etc). OPA (which
interprets Rego) does have a test feature:
https://www.openpolicyagent.org/docs/latest/policy-testing/
but this is not formal verification, it is case-by-case validation, though
it has a nice coverage output to allow you to see that you have reached all
lines of the policy. But it doesn't translate the Rego into FOL predicates
that can be input into a SAT or SMT solver, or model checker, so a Rego
policy still needs someone to construct a set of tests and mocks.  So I
would call this "validation", which is orthogonal and complimentary to
"verification".

That said, I think there are some interesting things we could do re:Rego:
1) formally verify that rego is correct.  maybe Rego has design or
implementation defects?
2) formally (via symbolic execution, static analysis) verify OPA is
correctly interpreting Rego
3) create an interpreter tool as part of OPA to translate Rego into FOL
predicates that can be fed into Z3 or other SAT/SMT solvers.

My initial goal start with the most basic language (essentially hard coded
Horn clauses) and build on that first. Clearly no one in practice is going
to take the time to build Horn clauses manually, so once I have a proof of
concept, I would love feedback on whether to use Datalog, or Rego, or some
other language to express the specification of policy that is more user
friendly.

Hope this helps, and I would welcome and/all feedback or participation!
Especially from users with real world use cases and problems to solve.
Send me any specific examples of problems you face with real world
kubernetes policies and I would gladly discuss how FV may or may not be a
fit!

**Comment by WeiZhang555** [comment]
@rficcaglia  Wow, thanks for your detailed information!  

> But it doesn't translate the Rego into FOL predicates
that can be input into an SAT or SMT solver, or model checker, so a Rego
policy still needs someone to construct a set of tests and mocks. 

I agree that this is a large missing part of OPA.  OPA is a great project and takes a lot of effort from people, its policy language is easy to use, but I think AWS Zelkova is more powerful than OPA since it's "verification" but OPA is "validation" just as you said. From this point, I'm not sure if we should enhance the OPA or open a new project and start from scratch.

My company(Ant Financial, one child corp of Alibaba) is a large financial company in China, we care a lot about security since "security" equals "money" to us 😆.  Also, we rely on K8s/Istio and other CNCF projects, so your project(Guardian) is an attractive and interesting idea for me.

I'm looking forward to your POC and will try to participate if there's something I can help 😄

**Comment by rficcaglia** [comment]
> I'm not sure if we should enhance the OPA or open a new project
well I think opa is trying to do two complimentary things very well, 1)
provide a language to express policy and 2) a lightweight high performance
agent to enforce policy. having worked with the team I think they have
their hands full working on those 2 priorities.  Adding formal verification
is interesting to them and we have discussed it (refer also to the
sig-security PR where Tim from OPA comments).  But I don't think it's
likely to be their primary focus.

Initially I thought about directly using FV with Rego and enhancing OPA,
but I decided a better first step is to do something completely separate
from basic principles to prove out the idea, then it could be applied to
Rego/OPA and other tools afterwards.  I think this mirrors how things
evolved at AWS, too, where Zelkova was a standalone tool but then gradually
incorporated into their commercial products (e.g. IAM, S3, GuardDuty, etc)
